### PR TITLE
[coq] [draft] Support for `coqffi`

### LINKED ISSUE #3995: Support coqffi in coq.theory stanza
Hi,

Lately, I have been working on  [`coqffi`](https://github.com/coq-community/coqffi), a tool to generate “FFI modules” to OCaml functions. It is my hope that `coqffi` will eventually be adopted by the Coq community, as a convenient way to mix Coq and OCaml code.

`coqffi` takes a `cmi` file as its input, and generate a Coq module. I have been using it with `dune` with a lot of success, as demonstrated by the `examples` directory of the repo (see the [dune file](https://github.com/coq-community/coqffi/blob/main/examples/src/dune)).

For instance,

```dune
(rule
  (target File.v)
  (action (run coqffi -finterface %{cmi:file} -o %{target})))
```

The fact that `dune` understands out of the box that `File.v` is a dependency of a theory declared with the `coq.theory` stanza is truly amazing. Thanks!

I was wondering if, when the time is right, it would be possible to add a dedicated field in `dune` to the `coq.theory` for instance, for `coqffi` to be more easily accessible to Coq users.

## Desired Behavior

Basically, the idea is to have a more straightforward way to generate FFI modules, for instance, `ffi_modules`.

## Example

```dune
(library MyLib)

(coq.theory
  (theory Test)
  (ffi_modules (MyLib.Test))
```

Which would means (1) generate the `Test.v` module from the `cmi` of the `Test` module from the `MyLib` library, and (2) add this module to the list of modules of the coq theory `Test`.

We could imagine being able to provide `coqffi` arguments, like

```dune
(coq.theory
  (theory Test)
  (ffi_modules ((MyLib.Test :flags -finterface))))
```

_________

So to wrap-up, do you think such a thing is possible? Is there a way for me to make it happen (which is not hacking `dune` to add the feature, but more from the `coqffi` side. The first obvious step is to publish a stable version of `coqffi` of course, but in addition to that, is there anything that comes to your minds?

Thanks for your feedback!

### PR DESCRIPTION
**Pull request by ejgallego**
This is a draft PR to support @lthms `coqffi` rules that allows for Coq to call OCaml libraries https://github.com/coq-community/coqffi . This has been written in collaboration with @lthms . Closes #3995

As of today, we add a `(ffi_modules ...)` field to `(coq.theory ...)` and the corresponding rules for `.cmi` to `.v` generation. This is a pretty small patch and highlights how to add a simple rule to Coq's mode.

There are some open issues and todos before we can consider this ready:

- _blocker_: fix code to locate a cmi from a lib + OCaml module name. As of today we just do a hack, however it doesn't work for wrapped modules; we couldn't find the right API for it. 
- syntax: there are several options, may need to specify flags, etc... so this is still work in progress
- documentation / changelog
- more tests [should be disabled by default until coqffi is stable]
- version check for `(coq.theory ...)`

### CODE REVIEW

File: src/dune_rules/coq_rules.ml [hunk 1]
```diff
@@ -394,6 +396,20 @@ let coq_modules_of_theory ~sctx lib =
...
   Coq_sources.library coq_sources ~name
 
+let ffi_rule ~dir ~coqffi ~expander (s : string) : Action.t Build.With_targets.t =
+  let ml_lib_name, coq_module_name = Coq_stanza.Theory.ffi_parse_name s in
+  let coq_v_name = Path.Build.relative dir (coq_module_name ^ ".v") in
+
+  (* Get the path to the cmi file *)
+  (* We need help here! The below fails with wrong dune *)
+  let expand_expr = String_with_vars.make_var Loc.none ~payload:ml_lib_name "cmi" in
+  let cmi_path = Expander.expand_path expander expand_expr in
```
Comment (ejgallego) [inline src/dune_rules/coq_rules.ml:406]: This doesn't work, we also tried:
```ocaml
  let cmi_path =
    let open Result.O in
    let* ml_lib = Lib.DB.resolve lib_db (Loc.none, Lib_name.of_string ml_lib_name) in
    let info = Lib.info ml_lib in
    let obj_dir = Obj_dir.public_cmi_dir (Lib_info.obj_dir info) in
    Result.Ok (Path.relative obj_dir (coq_module_name ^ ".cmi"))
```
which works but it doesn't compute the right cmi path for wrapped libs.

**Comment by ejgallego** [comment]
Updated with a fresh attempt to locate a cmi path, but unfortunately it only works for local libs.

**Comment by lthms** [comment]
> Updated with a fresh attempt to locate a cmi path, but unfortunately it only works for local libs.

This is already a pretty great improvement, congrats!

**Comment by ejgallego** [comment]
Adding @rgrinberg 's comments on how to move forward with this:

> The correct way is to find the `Obj_dir.t` that corresponds to this library and then use it to resolve the module. You can obtain the module by name by getting the `Modules.t` of the library.

> For external libraries, the `Modules.t` should be available straight from`dune-package`.

**Comment by lthms** [comment]
Thanks @ejgallego (and @rgrinberg ) for the heads-up. I have tagged a first beta release for `coqffi`, so I will have some time now to work on this. I hope I will be able to have a first clean version for modules from local libraries later this week!

**Comment by ejgallego** [comment]
Hi @lthms , I'm closing this PR due to inactivity, but I'd love to reopen and finish it if you still find it useful.

**Comment by lthms** [comment]
Yeah, sorry about that. I’ve been focused on other things. Yet, `coqffi` has become way more stable now, and I think that as soon as I release the first stable version, I will give this another try.

Thanks again a lot for taking the time to guide me the other time.

**Comment by ejgallego** [comment]
Great to hear, looking forward to completing this, and to using coqffi

**Pull request closed without merging**